[1.16] ForgeEventFactory#canCreateFluidSource reintroduced

Edivad99 opened the pull request:
I reintroduced the ForgeEventFactory#canCreateFluidSource event, this allows you to set the water as not infinite.
![2020-07-27_17 23 19](https://user-images.githubusercontent.com/25778604/88566475-591b6380-d036-11ea-82e3-644df472181c.png)
![2020-07-27_17 22 44](https://user-images.githubusercontent.com/25778604/88566489-5d478100-d036-11ea-89b7-a0fb74d1a6c5.png)
This should close this issue #6987 
If I have to change something, just tell me, since it's my first PR for Forge

Edivad99 committed:
ForgeEventFactory#canCreateFluidSource reintroduced

Edivad99 committed:
Merge branch '1.16.x' into 1.16.x

Edivad99 committed:
Update mod name

ichttt commented on src/test/java/net/minecraftforge/debug/fluid/FiniteWaterTest.java:
```diff
@@ -0,0 +1,41 @@
+
+@Mod("finite_water_test")
+@Mod.EventBusSubscriber()
+public class FiniteWaterTest {
```
Should probably be disabled by default. Also, please keep you curly brackets style at least consistent within on class and move it to a new line.

ichttt commented on src/test/java/net/minecraftforge/debug/fluid/FiniteWaterTest.java:
```diff
@@ -0,0 +1,41 @@
+    {
+        BlockState state = event.getState();
+        FluidState fluidState = state.getFluidState();
+        if(fluidState.getFluid().isEquivalentTo(Fluids.WATER))
```
spaces between if and bracket please

ichttt commented on patches/minecraft/net/minecraft/fluid/FlowingFluid.java.patch:
```diff
@@ -0,0 +1,11 @@
+          FluidState fluidstate = blockstate.func_204520_s();
+          if (fluidstate.func_206886_c().func_207187_a(this) && this.func_212751_a(direction, p_205576_1_, p_205576_2_, p_205576_3_, blockpos, blockstate)) {
+-            if (fluidstate.func_206889_d()) {
++            if (fluidstate.func_206889_d() && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource((World) p_205576_1_, p_205576_2_, blockstate, this.func_205579_d())) {
```
This is not the correct place to insert the patch. Move it below to where `if (this.canSourcesMultiply() && j >= 2)` is checked, and replace it with `j >= 2 && if (j >= 2 && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(worldIn, pos, blockStateIn, canSourcesMultiply()))`, so you can also use the event the other way around (making fluid sources for blocks that normally don't make them).
It would also be better if we could just change the method to accept IWorldReader instead of casting to World.

Edivad99 committed:
Merge pull request #1 from MinecraftForge/1.16.x

Make license toml entry optional during 1.16.1

Edivad99 committed:
Adjusted based on recommendations

Edivad99 committed:
Update FiniteWaterTest.java

Edivad99 commented:
@ichttt I did some tests, as you say it is wrong because if I place 3 buckets of water in a row, I get an infinite source of water. While where I had put the code earlier, if I took water from the center the infinite source disappeared.
I think I should go back to the old changes.

Edivad99 commented:
I add that I don't think you can do the opposite (like creating infinite lava) as there is this code.
`            if (fluidstate.isSource()) {
               ++j;
            }
`

Edivad99 committed:
Now we can create infinite lava sources

Edivad99 commented:
With the last commit, everything works and infinite lava sources can be created

Edivad99 commented:
ichttt, I wanted to know what you think of the last commit, I tried to put it as you say but it doesn't work. Did you plan on doing this??
```
if (fluidstate.getFluid().isEquivalentTo(this) && this.doesSideHaveHoles(direction, worldIn, pos, blockStateIn, blockpos, blockstate)) {
            if (fluidstate.isSource()) {
               ++j;
            }

            i = Math.max(i, fluidstate.getLevel());
         }
      }

if (j >= 2 && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(worldIn, pos, blockStateIn, this.canSourcesMultiply())) {
```

Edivad99 commented:
ichttt, I wanted to know what you think of the last commit, I tried to put it as you say but it doesn't work. Did you plan on doing this?
```
    FluidState fluidstate = blockstate.getFluidState();
    if (fluidstate.getFluid().isEquivalentTo(this) && this.doesSideHaveHoles(direction, worldIn, pos, blockStateIn, blockpos, blockstate)) {
        if (fluidstate.isSource()) {
            ++j;
        }

        i = Math.max(i, fluidstate.getLevel());
    }
}

if (j >= 2 && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(worldIn, pos, blockStateIn, this.canSourcesMultiply())) {
```

Edivad99 commented:
As I wrote above, your idea works halfway, if I place the water in position 1 and 3, correctly it does not create infinite water in position 2. But if I take a bucket of water and put it in position 2, I get an infinite source of water.
+-+-+-+
+1|2|3+
+-+-+-+

Edivad99 committed:
Merge branch '1.16.x' into 1.16.x

Edivad99 commented:
@ichttt Is that okay?  I wish this event was implemented quickly because I need it for my mod.

diesieben07 commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -661,7 +662,7 @@ public static LootTable loadLootTable(ResourceLocation name, LootTable table, Lo
     }
 
-    public static boolean canCreateFluidSource(World world, BlockPos pos, BlockState state, boolean def)
+    public static boolean canCreateFluidSource(IWorldReader world, BlockPos pos, BlockState state, boolean def)
```
This is a binary-breaking change and won't be accepted for 1.16.

diesieben07 commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -281,9 +282,9 @@ public boolean getForceRedstoneUpdate()
     public static class CreateFluidSourceEvent extends BlockEvent
     {
-        public CreateFluidSourceEvent(World world, BlockPos pos, BlockState state)
+        public CreateFluidSourceEvent(IWorldReader world, BlockPos pos, BlockState state)
```
Binary breaking change

diesieben07 commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -281,9 +282,9 @@ public boolean getForceRedstoneUpdate()
+        public CreateFluidSourceEvent(IWorldReader world, BlockPos pos, BlockState state)
         {
-            super(world, pos, state);
+            super((IWorld) world, pos, state);
```
You can't just blindly cast things.

Edivad99 replied:
What do you think about it? `super(world instanceof IWorld ? (IWorld) world : null, pos, state);`

diesieben07 commented on src/test/java/net/minecraftforge/debug/fluid/FiniteWaterTest.java:
```diff
@@ -0,0 +1,50 @@
+    @SubscribeEvent
+    public static void handleFiniteWaterSource(BlockEvent.CreateFluidSourceEvent event)
+    {
+        if (ENABLED) {
+            BlockState state = event.getState();
+            FluidState fluidState = state.getFluidState();
+            if (fluidState.getFluid().isEquivalentTo(Fluids.WATER)) {
+                event.setResult(Event.Result.DENY);
+            }
+            else if (fluidState.getFluid().isEquivalentTo(Fluids.LAVA)) {
+                event.setResult(Event.Result.ALLOW);
+            }
```
Formatting (braces on newlines).

Edivad99 committed:
Applied the suggestions of diesieben07

Edivad99 committed:
Merge branch '1.16.x' of https://github.com/Edivad99/MinecraftForge into 1.16.x

diesieben07 commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -282,9 +282,15 @@ public boolean getForceRedstoneUpdate()
         public CreateFluidSourceEvent(IWorldReader world, BlockPos pos, BlockState state)
         {
-            super((IWorld) world, pos, state);
+            super(world instanceof IWorld ? (IWorld) world : null, pos, state);
```
So if it's not an `IWorld` you can't access it at all? That's not helpful...

Edivad99 committed:
Changed the type of parameters in FlowingFluid.java from IWorldReader to IWorld

Edivad99 committed:
Update FiniteWaterTest.java

Edivad99 commented:
@diesieben07 I changed the type of parameters in FlowingFluid.java from IWorldReader to IWorld. What do you think?

diesieben07 commented:
You can't do that either, as some people might be subclassing that class. As such that's still binary breaking.

Edivad99 commented:
So I can't do anything? ahahah

Edivad99 commented:
But being Forge 1.16.1 still in beta, can't we do the same?

diesieben07 commented:
1.16 is technically still in "beta", but it has been so long now that even without an RB, breaking changes are not allowed anymore.

Edivad99 commented:
So how should I move?
I can't touch any method ...

TheCurle commented:
1.16.2 is going to be a binary break. Once it lands, rebase your changes and we might be able to implement it then.

Edivad99 commented:
Okay, so let's see if they release it this week and I'll update this PR.

Edivad99 commented:
Do I leave the deprecated methods in 1.16.2?

diesieben07 commented:
Since 1.16.2 will be binary-breaking, you can remove them.

Edivad99 committed:
Update FiniteWaterTest.java

Edivad99 committed:
Merge branch '1.16.x' of https://github.com/MinecraftForge/MinecraftForge into MinecraftForge-1.16.x

Edivad99 committed:
Merge branch 'MinecraftForge-1.16.x' into 1.16.x

Edivad99 committed:
Now it works for 1.16.2

Edivad99 commented:
@diesieben07 done now also works for Minecraft 1.16.2

Edivad99 committed:
Update BlockEvent.java

Edivad99 committed:
Update ForgeEventFactory.java

Edivad99 commented:
https://discord.com/channels/313125603924639766/725850371834118214/745547827693879298

Edivad99 committed:
CreateFluidSourceEvent inherits from Event

based on discord suggestions https://discord.com/channels/313125603924639766/725850371834118214/745547827693879298
